waspc/src/Wasp/Analyzer/Parser/Parser.hs

```diff
@@ -0,0 +1,309 @@
+{-# LANGUAGE LambdaCase #-}
+
+module Wasp.Analyzer.Parser.Parser
+  ( -- * AST to CST conversion
+
+    -- | This module takes @["SyntaxNode"]@ produced by 'Wasp.Analyzer.Parser.ConcreteParser'
+    -- and converts it into an abstract syntax tree.
+    parseStatements,
+    parseExpression,
+  )
+where
+
+import Control.Arrow (Arrow (first))
+import Control.Monad.Except (Except, runExcept, throwError)
+import Control.Monad.State.Strict (StateT, evalStateT, gets, modify)
+import Data.Maybe (catMaybes)
+import Wasp.Analyzer.Parser.AST (AST, Expr, ExtImportName, Identifier, Stmt)
+import qualified Wasp.Analyzer.Parser.AST as AST
+import Wasp.Analyzer.Parser.ConcreteParser.CST (SyntaxKind, SyntaxNode (SyntaxNode))
+import qualified Wasp.Analyzer.Parser.ConcreteParser.CST as S
+import Wasp.Analyzer.Parser.Ctx (Ctx (Ctx), WithCtx (WithCtx), ctxFromRgn)
+import Wasp.Analyzer.Parser.ParseError
+import Wasp.Analyzer.Parser.SourcePosition (SourcePosition (SourcePosition))
+import Wasp.Analyzer.Parser.SourceRegion (SourceRegion (SourceRegion))
+import qualified Wasp.Analyzer.Parser.Token as T
+
+data ParseState = ParseState
+  { pstateLine :: !Int,
+    pstateColumn :: !Int,
+    pstateRemainingSource :: String
+  }
+
+pstatePos :: ParseState -> SourcePosition
+pstatePos s = SourcePosition (pstateLine s) (pstateColumn s)
+
+type ParserM a = StateT ParseState (Except ParseError) a
+
+-- | An operation that turns some syntax nodes into a value and outputs the
+-- remaining nodes.
+type Action a = [SyntaxNode] -> ParserM (a, [SyntaxNode])
```

Comment:
bump

Reply:
I agree that `Action` is a pretty bad name. I'll try to clarify the different concepts going on:

- `SyntaxNode -> ParserM a`: Make a value from and use a single node
- `[SyntaxNode] -> ParserM a`: Makes a value from a list of nodes, and guarantees that it will use every `[SyntaxNode]` in that list
- `Action a`: Makes a value from a list of nodes, but it will also return nodes that it doesn't use

Using a node just means that it's width is used to increase the offset.

`coerce` is used because the CST is being "coerced" into an AST. But maybe that confuses things by adding in another word, I think `coerce` can be search-and-replaced to `parse` and names will still make sense.

I gave only `Action` its own name because its the only one of three classes of parsing functions that have complicated composition rules (the reason `sequence2`, etc., exist).

Perhaps we could name them something like `NodeParser`, `NodesParser`, `PartialNodesParser`?

Reply:
Good analysis, and yes I like those names!
Could be `SingleNodeParser`, `NodesParser` and `NodesPartialParser`.